File: decoding-string.py
# Brute Force Solution - Using a stack
class Solution(object):
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        stack = []
        for i in range(len(s)):
            if s[i] == ']':
                s1 = []
                while stack[-1] != '[':
                    s1.append(stack.pop())
                stack.pop()
                k = 0
                base = 1
                while stack and stack[-1].isnumeric():
                    k = k + int(stack.pop()) * base
                    base = base * 10
                
                while k != 0:
                    for j in reversed(range(len(s1))):
                        stack.append(s1[j])
                    k -= 1
                
            else:
                stack.append(s[i])
        
        return ''.join(stack)

File: decodingString.java
// Time Complexity : (N) N=length of string
// Space Complexity : O(2N) 2 stacks of size N
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/*
    Add operations into stack and po them whenwe see closing brackets   
*/


// Your code here along with comments explaining your approach
class Solution {
    public String decodeString(String s) {
        if(s == null || s.length() == 0) return s;
        int num = 0; StringBuilder currStr = new StringBuilder();
        Stack<Integer> numStack = new Stack<>();
        Stack<StringBuilder> strStack = new Stack<>();
        
        for(int i=0;i<s.length();i++) {
            char c = s.charAt(i);
            
            if(Character.isDigit(c)) {
                num = num * 10 + c - '0';
            } else if(c == '[') {
                numStack.push(num);
                strStack.push(currStr);
                num = 0;
                currStr = new StringBuilder();
            } else if(c == ']') {
                int times = numStack.pop();
                StringBuilder newStr = new StringBuilder();
                for(int k=0;k<times;k++) {
                    newStr.append(currStr);
                }
                StringBuilder parent = strStack.pop();
                currStr = parent.append(newStr);
            } else {
                currStr.append(c);
            }
        }
        return currStr.toString();
    }
}
